Configurable tools

damphyr opened the pull request:
Pulls out the tool configuration in a YAML file under config.

Also makes all constant values (CACHE_DIR, ZIP_EXE etc.) overridable through environment variables (which you can use on the rake command line)

Updates all software to the latest versions.

Commit:
YAML configuration with updated tool versions

Commit:
updated gem versions

Commit:
download and unpack uses the YAML config

damphyr commented:
Not ready yet, PR opened to be able to clarify a few things

damphyr commented:
Q1: What are you doing with that 'includes' parameter?

Commit:
chef, atom and vagrant plugins read plugin list and versions from the YAML configuration. All chef, vagrant and atom extras are conditional on the presence of the tool in the configuration

Commit:
pass the tool config

damphyr commented:
That Appveyor build is going to block forever because it's asking for the git name and email

Commit:
not an Array of Hashes, just a Hash with name=<version

damphyr commented:
Torben I need your input here, there's a few things that do not work due to environment scripts:
- Asking for the name and email for git in set-env blocks all builds.
- Atom plugin installation fails.

On the other hand I've restructured most of the Rakefile so that we can use it as a basis for all devpacks and handle exceptions (like the double unpack for Chef).

Basically I broke the kitchen and took home the oven to bake my own cookies.

tknerr commented on Rakefile:
```diff
@@ -1,42 +1,76 @@
+  def target_dir
+    ddir=ENV["TARGET_DIR"]
+    ddir||=File.join(base_dir,"target")
+    File.expand_path(ddir)
```
how about `File.expand_path(ENV["TARGET_DIR"] || File.join(base_dir, "target")` instead?

for my own taste a bit more readable as a one-liner, but not being picky here

tknerr commented on Rakefile:
```diff
@@ -110,114 +144,96 @@ def purge_atom_plugins_with_insanely_long_path
+  #chefdk install package contains a zip file
+  extra_pkg="#{target_dir}/chefdk/opscode/chefdk.zip"
+  tgt_dir="#{target_dir}/chefdk/opscode/"
+  unpack(extra_pkg, tgt_dir, includes = [])
```
need the `extra_pkg` and `tgt_dir` vars here? I'd rather inline them here...

tknerr commented on Rakefile:
```diff
@@ -1,42 +1,76 @@
 
 desc 'downloads required resources and builds the devpack binary'
 task :build => :clean do
+  tools_config=YAML.load(File.read("#{base_dir}/config/tools.yaml"))
```
yes, I like that! :-)

tknerr commented on Rakefile:
```diff
@@ -1,42 +1,76 @@
-  move_chefdk
-  fix_chefdk
+  download_tools(tools_config)
+  if tools_config.keys.include?("chefdk")
```
wonder if we could make this a bit more generic and pluggable. not necessarily for bills-kitchen here, but if we want to use that as a minimal template for future devpacks

tknerr replied:
All the `if`s in the :build task I mean...

tknerr commented on Rakefile:
```diff
@@ -110,114 +144,96 @@ def purge_atom_plugins_with_insanely_long_path
 end
 
-def install_knife_plugins
+def install_knife_plugins tool_config
```
nice one

tknerr commented on Rakefile:
```diff
@@ -110,114 +144,96 @@ def purge_atom_plugins_with_insanely_long_path
 end
 
-def install_vagrant_plugins
+def install_vagrant_plugins tool_config
```
nice one

tknerr commented on Rakefile:
```diff
@@ -110,114 +144,96 @@ def purge_atom_plugins_with_insanely_long_path
 end
 
-def install_atom_plugins
+def install_atom_plugins tool_config
```
nice one, too.

tknerr reviewed (commented):
LGTM, but no idea why the appveyor build is failing yet. Suspect this is a side effect of the tool updates. Will have a look....

Thanks for the improvements!

tknerr committed:
correct some paths and restore the original logic in move_chefdk

both the build_dir and cache_dir were nested under target_dir originally. restoring that for now...

tknerr committed:
adapt tool name to chef_dk as expected in the Rakefile

(earlier this was cdk, hope it won't hit the max path length issue with chef_dk...)

tknerr committed:
restore the original tool names so they match the paths defined in set-env.bat

tknerr committed:
fix vagrant plugin installation

tknerr commented:
@damphyr while I believe we could get this to work / pass the appveyor build again, I'm not sure what to do with it.

Instead of doing something half-assed, I'd rather focus on making https://github.com/Zuehlke/z-devpack reusable / modular enough (see https://github.com/Zuehlke/z-devpack/issues/1) so it can be the new basis for bills-kitchen.

What do you think?

damphyr commented:
What I think is : https://github.com/Zuehlke/z-devpack :)

damphyr commented:
Do we scrap this or should I bring it up to date?

tknerr commented:
I should probably add some large banner stating that this is unmaintained (ever since I switched to macOS about 1.5 years ago...)

I'm currently maintaining https://github.com/tknerr/linus-kitchen/ instead, which is a Linux Developer VM variant with the same intent / goal as Bill's Kitchen.

So yes, I'm in favor of ditching this, unless someone wants to continue to maintain it
